discord/commands/mixins.py

```diff
@@ -0,0 +1,895 @@
+        """
+        if not asyncio.iscoroutinefunction(coro):
+            raise TypeError("The post-invoke hook must be a coroutine.")
+
+        self._after_invoke = coro
+        return coro
+
+    async def can_run(self, ctx: ContextT) -> bool:
+        """|coro|
+
+        Checks if the command can be executed by checking all the predicates
+        inside the :attr:`~Command.checks` attribute. This also checks whether the
+        command is disabled.
+
+        .. versionchanged:: 1.3
+            Checks whether the command is disabled or not
+
+        Parameters
+        -----------
+        ctx: :class:`.Context`
+            The ctx of the command currently being invoked.
+
+        Raises
+        -------
+        :class:`CommandError`
+            Any command error that was raised during a check call will be propagated
+            by this function.
+
+        Returns
+        --------
+        :class:`bool`
+            A boolean indicating if the command can be invoked.
+        """
+        if not self.enabled:
+            raise DisabledCommand(f"{self.name} command is disabled")
+
+        original = ctx.command
+        ctx.command = self
+
+        try:
+            if not await ctx.bot.can_run(ctx):
+                raise CheckFailure(f"The global check functions for command {self.qualified_name} failed.")
+
+            if (cog := self.cog) and (local_check := cog._get_overridden_method(cog.cog_check)):
+                ret = await utils.maybe_coroutine(local_check, ctx)
+                if not ret:
+                    return False
+
+            predicates = self.checks
+            if not predicates:
+                # since we have no checks, then we just return True.
+                return True
+
+            return await utils.async_all(predicate(ctx) for predicate in predicates)
+        finally:
+            ctx.command = original
+
+    def add_check(self, func: Check) -> None:
+        """Adds a check to the command.
+
+        This is the non-decorator interface to :func:`.check`.
+
+        Parameters
+        -----------
+        func: Callable
+            The function that will be used as a check.
+        """
+
+        self.checks.append(func)
+
+    def remove_check(self, func: Check) -> None:
+        """Removes a check from the command.
+
+        This function is idempotent and will not raise an exception
+        if the function is not in the command's checks.
+
+        Parameters
+        -----------
+        func: Callable
+            The function to remove from the checks.
+        """
+
+        try:
+            self.checks.remove(func)
+        except ValueError:
+            pass
+
+    def copy(self):
+        """Creates a copy of this command.
+
+        Returns
+        --------
+        :class:`Invokable`
+            A new instance of this command.
+        """
+        ret = self.__class__(self.callback, **self.__original_kwargs__)
+        return self._ensure_assignment_on_copy(ret)
```

Comment:
Is there a reason why `copy.copy` isn't used here?

Reply:
Apparently it was just implemented like that (since [here](https://github.com/Pycord-Development/pycord/commit/caf3d17d4aa3ce45435bb96ea6d99317bfef8618)), probably because of some technical reasons. I tested it out a bit but I think it should be it's own PR if it ends up working.

Reply:
Why should it be in its own PR?

Reply:
After a bit more testing and such, I think the reason for a custom copy method is for setting cog-wide attributes via parameters with `_update_copy`
https://github.com/Pycord-Development/pycord/blob/7c8cd56c4a9d1ae7900578db351c24e508776de0/discord/cog.py#L238
https://github.com/Pycord-Development/pycord/blob/7c8cd56c4a9d1ae7900578db351c24e508776de0/discord/commands/core.py#L965-L972
I haven't found a way to do the same with copy or deepcopy so far

Reply:
> In order for a class to define its own copy implementation, it can define special methods `__copy__()` and `__deepcopy__()`.
https://docs.python.org/3.8/library/copy.html

Reply:
🤔